File: LeetCode/Hard/3333-Find_The_Original_Typed_String_II/3333-FindTheOriginalTypedStringII.cpp
static const int MOD = 1000000007;

class Solution {
public:
    int possibleStringCount(string word, int k) {
        std::vector<int> groups;
        int n = word.size();

        int count = 1;

        for (int i = 1; i < n; i++) {
            if (word[i - 1] == word[i]) {
                count++;
            } else {
                groups.push_back(count);
                count = 1;
            }
        }

        groups.push_back(count);

        long long total = 1;

        for (int size: groups) {
            total = (total * size) % MOD;
        }

        if (groups.size() >= k)
            return total;

        std::vector<long long> prefix_sum(k, 1);
        std::vector<long long> dp;

        for (int size : groups) {
            dp = std::vector<long long>(k, 0);

            for (int length = 1; length < k; length++) {
                dp[length] = prefix_sum[length - 1];

                if (length - size - 1 >= 0)
                    dp[length] = (dp[length] - prefix_sum[length - size - 1] + MOD) % MOD;
            }

            prefix_sum = std::vector<long long>(k, 0);

            for (int i = 1; i < k; i++) {
                prefix_sum[i] = (prefix_sum[i - 1] + dp[i]) % MOD;
            }
        }

        return (total - prefix_sum[k - 1] + MOD) % MOD;
    }
};

File: LeetCode/Hard/3333-Find_The_Original_Typed_String_II/3333-FindTheOriginalTypedStringII.py
class Solution:
    def possibleStringCount(self, word: str, k: int) -> int:
        MOD = 10**9 + 7
        groups = []
        n = len(word)

        count = 1

        for i in range(1, n):
            if word[i - 1] == word[i]:
                count += 1
            else:
                groups.append(count)
                count = 1

        groups.append(count)

        total = 1

        for size in groups:
            total *= size % MOD

        if len(groups) >= k:
            return total

        prefix_sum = [1] * k

        for size in groups:
            dp = [0] * k

            for length in range(1, k):
                dp[length] = prefix_sum[length - 1]

                if length - size - 1 >= 0:
                    dp[length] -= prefix_sum[length - size - 1] % MOD

            prefix_sum = [0] * k

            for i in range(1, k):
                prefix_sum[i] = prefix_sum[i - 1] + dp[i]

        return (total - prefix_sum[k - 1]) % MOD


a = Solution()
print(a.possibleStringCount(word="aabbccdd", k=7))
print(a.possibleStringCount(word="aabbccdd", k=8))
print(a.possibleStringCount(word="aaabbb", k=3))
